1-Draft/RFCNNNN-Supports-Parallel.md

```diff
@@ -0,0 +1,316 @@
+        #               parameters are used in the invocation.
+        # [-ThreadLimit <int>] : An int used to define the maximum number of
+        #                        threads to use during the parallel command
+        #                        execution (default value depends on system
+        #                        configuration)
+        # [-TimeoutSecs <int>] : An int used to define the maximum time allowed
+        #                        for all threads to complete. This parameter
+        #                        cannot be used in conjunction with the next
+        #                        two parameters.
+
+        # NOTE! [-AsJob] and [-JobName <string>] are intentionally excluded
+        # from this list because that work must be done external to the
+        # pipeline. The reasoning for this decision is explained in the
+        # "Alternate Proposals and Considerations" section below, under the
+        # subheading of "Running a parallelized command asynchronously".
+    )
+    begin {
+        # No changes here, this would be invoked synchronously in the current
+        # session.
+    }
+    process {
+        # NEW! The way this is invoked depends on whether or not the parameter
+        # set that was used supports parallelization or not.
+        #
+        # For a parameter set that does not support parallelization, this would
+        # be invoked synchronously.
+        #
+        # For a parameter set that supports parallelization, this would be
+        # invoked in a thread job, and the command would wait for all thread
+        # jobs to complete before continuing with end (and dispose) blocks.
+
+        # NEW! Unlike Start-Job commands where you must use the "using:" prefix
+        # on variables to access variables that are defined outside of the
+        # scope of the script block, all variables that are defined in the
+        # current scope would be accessible when the command is invoked with
+        # parallel processing without having to use the "using:" prefix. This
+        # is a smarter design for working with variables from other runspaces,
+        # and allows for easier transition of a command that does not support
+        # parallel processing into a command that does support parallel
+        # processing. The "using:" qualifier should only be used when you want
+        # to reference a variable that is outside of the scope of the local
+        # function.
+
+        # NEW! "using:" does not allow for transport of script blocks. This was
+        # done as a security measure (see [PowerShell Issue 9703](https://github.com/PowerShell/PowerShell/issues/9703) for details);
+        # however, there are valid cases where script blocks must be
+        # transportable into jobs, and this feature is one such case. Any
+        # variables that contain script blocks (i.e. parameters, variables
+        # defined in a begin block, etc.) must be made available to the thread
```

Comment:
Can you elaborate on how you envision this working?

Currently scriptblocks are bound to either the "global" or a module's SessionState.  Each scriptblock can read and alter the SessionState to which it is bound when it is invoked.  This works fine in a single-threaded scenario because only one scriptblock at a time can alter a given SessionState.  With a second thread, however, two threads could be accessing the same SessionState and currently results in a crash.  That is the topic of PowerShell/PowerShell#7626 which does not seem to have an obvious resolution.